### PR TITLE
feat(rm-001): ReleaseManager MVP — 6 workflows + 3 scripts + tests

### PR DESCRIPTION
**Pull request by mellanon**
Implements **Phase 3 of [meta-factory#390](https://github.com/the-metafactory/meta-factory/issues/390)** for [release-manager#1](https://github.com/the-metafactory/release-manager/issues/1) — the skill bundle Forge composes for end-to-end releases.

## Summary

- 6 workflow MDs in `skill/Workflows/` — each follows the Action / Verify / Anti-pattern triplet contract
- 3 bun-runnable scripts in `scripts/` for the moving parts a workflow needs
- 69 tests across 3 files (target was ≥20)
- Day-one scope: **grove only**. `// TODO(v0.2): generalize for non-grove repos` markers flag every grove-specific path. Generalisation is a Phase G follow-up — do not add per-repo branching here yet.
- `ScaffoldInstance` is intentionally absent from this bundle; it lives in `AgentState`.

## Workflows shipped

| #  | Workflow         | Day-one behaviour                                                                          |
|----|------------------|--------------------------------------------------------------------------------------------|
| 1  | BumpVersion      | Commit-prefix scan → patch/minor/major; edit `arc-manifest.yaml`; commit on default branch |
| 2  | CutRelease       | `gh release create` with `--generate-notes --notes-start-tag v<last>`                      |
| 3  | TrustGateCheck   | Walk Phase 0 gate table from `compass/sops/release-checklist.md`; halt on first fail       |
| 4  | DeployStaged     | dev first → prod second, **separate workflow runs**; prod refuses without 4h-fresh dev     |
| 5  | Rollback        | Checkout prior tag, redeploy prod, record cause in events pipeline                          |
| 6  | Announce         | Structured release note → `#releases` + tracking issue; always **last** step                |

## Scripts shipped

| Script | Purpose |
|---|---|
| `list-releases-since-tag.ts` | Enumerate commits + merged PRs since a tag → JSON |
| `prepare-changelog.ts` | Group PRs by conventional-commit prefix → release-note markdown |
| `check-gate-table.ts` | Parse Phase 0 gate table; run rows via `bash`; emit `{gates, allPassed}` JSON. Flags: `--milestone`, `--dry-run`, `--checklist` |

## Verify

- `bunx tsc --noEmit` — clean
- `bun test` — 69 pass / 0 fail / 101 expects
- `bun scripts/list-releases-since-tag.ts ~/Developer/grove v0.22.3` — sane JSON with commits + PRs
- `bun scripts/prepare-changelog.ts ~/Developer/grove v0.24.5` — non-empty grouped markdown
- `bun scripts/check-gate-table.ts --dry-run --milestone S2-22` — 5 rows from live release-checklist.md

## Review focus

Per the assignment brief, please weight on:

1. **Gate-table parsing** in `scripts/check-gate-table.ts` (`parseGateTable`, `isExecutable`, the Phase 1 boundary detection that prevents bleed past Phase 0).
2. **Prod-deploy gate logic** in `skill/Workflows/DeployStaged.md` — specifically the requirement that prod refuses unless a `system.deploy.dev` event exists within 4h, and that prod must always be a separate workflow run.

## References

- Closes part of [release-manager#1](https://github.com/the-metafactory/release-manager/issues/1) (Phase 3)
- Phase 3 of [meta-factory#390](https://github.com/the-metafactory/meta-factory/issues/390)
- Forge design: `forge/design/forge.md` §"ReleaseManager bundle (what it ships)"
- Compass SOPs walked: `release-checklist.md`, `versioning.md`, `deployment.md`

## Test plan

- [ ] `bunx tsc --noEmit` clean
- [ ] `bun test` — 69 pass
- [ ] Manual smoke: `bun scripts/list-releases-since-tag.ts ~/Developer/grove v0.22.3` returns sane JSON
- [ ] Manual smoke: `bun scripts/prepare-changelog.ts ~/Developer/grove v0.24.5` returns non-empty markdown
- [ ] Manual smoke: `bun scripts/check-gate-table.ts --dry-run --milestone S2-22` returns the gate-1 rows
- [ ] Workflow MDs all parse with the action/verify/anti-pattern shape (covered by `workflow-shape.test.ts`)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/check-gate-table.ts [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    process.stdout.write(helpText() + "\n");
+    return 0;
+  }
+  if (!existsSync(args.checklist)) {
+    process.stderr.write(`check-gate-table: checklist not found: ${args.checklist}\n`);
+    return 1;
+  }
+  const markdown = readFileSync(args.checklist, "utf8");
+  const rows = parseGateTable(markdown);
+  if (rows.length === 0) {
```
Comment (mellanon) [inline scripts/check-gate-table.ts:498]: **[blocker/security/correctness]** `runChecklist` returns `allPassed: true` when there are zero rows.

If `parseGateTable` returns an empty array for any reason — checklist file moves, structure changes, milestone filter matches no rows, parser breaks on a future format — line 244 emits a stderr warning but `runChecklist([])` then runs `results.every(...)` on an empty array, which returns `true`. The script exits 0 with `{gates: [], allPassed: true}`.

`DeployStaged.md` pre-flight reads exactly that signal: *"TrustGateCheck (if applicable to this milestone) has reported `allPassed: true`."* So an unparseable gate table = silent green light to prod. The single most important rule of this workflow is *halt on first fail*; right now an empty/missing table is the same as a green table.

**Fix:** zero parsed rows must be a hard fail, not a warning. Either return `allPassed: false` when `rows.length === 0`, or have `main()` exit non-zero before `runChecklist` ever runs. Add a test asserting `{gates: [], allPassed: false}` for the empty case.

File: scripts/check-gate-table.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+export function runChecklist(
+  rows: GateRow[],
+  opts: RunChecklistOptions = {},
+): CheckGateReport {
+  const filtered = opts.milestone
+    ? rows.filter((r) => r.milestone === opts.milestone)
+    : rows;
+  const results = filtered.map((row) => runGateRow(row, opts));
+  const allPassed = results.every((r) => r.status === "pass" || r.status === "skipped");
```
Comment (mellanon) [inline scripts/check-gate-table.ts]: **[major/correctness]** `skipped` rows count toward `allPassed: true`.

Line 180 treats `pass` and `skipped` identically for the green-light flag. A milestone whose gate table is entirely manual cells ("Manual review by sponsor", "Screenshot", etc.) returns `allPassed: true` with the same automated signal as a milestone where every executable command actually returned 0.

`DeployStaged.md` pre-flight reads only `allPassed`. So an all-manual milestone passes pre-flight without anyone confirming the manual gates. The `TrustGateCheck.md` SOP says "the operator has acknowledged the gate report" — but that's a human contract, not a machine-checked one.

**Fix:** distinguish three states. Either:
- `allPassed: true` only when there is at least one executed row and zero failures (skipped excluded), plus a separate `pendingManual: true` flag the workflow consumer must resolve, OR
- a flag like `--require-all-executed` that flips skipped → fail.

The current shape is the least safe option: skipped is silently green.

File: scripts/check-gate-table.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+  }
+  if (!isExecutable(row.verify)) {
+    return {
+      ...row,
+      status: "skipped",
+      output: "verify cell is not an executable command (manual gate)",
+    };
+  }
+  const command = unwrapBacktickCommand(row.verify);
+  const result = spawnSync("bash", ["-c", command], {
```
Comment (mellanon) [inline scripts/check-gate-table.ts]: **[major/hardening]** `bash -c` execution of attacker-controllable markdown content.

`runGateRow` runs `spawnSync("bash", ["-c", command])` where `command` comes from a cell in `compass/sops/release-checklist.md`. Anyone who can edit that file can run arbitrary shell on the release operator's machine. Today compass is internal/trusted, so the exploit surface is small — but the script is now *the* mechanism that turns markdown into shell on every release.

This is the kind of trust boundary that needs to be loud, not implicit:

- Document the trust model at the top of the file: "This script executes shell commands embedded in `release-checklist.md`. Treat that file as code; only commits from trusted reviewers are allowed."
- Optional defence-in-depth: enforce a regex allowlist for verify cells (e.g. only allow commands matching a known prefix list — `gh`, `git`, `curl`, etc.), or refuse cells containing `;`, `|`, `&&`, `$(` unless an opt-in flag is passed.

The current 30s timeout is the only mitigation, and it doesn't help against `rm -rf ~ \&` — that's done in 30ms.

File: scripts/list-releases-since-tag.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    if (ghResult.status === 0 && ghResult.stdout) {
+      try {
+        prs = parseGhPrList(ghResult.stdout);
+      } catch (err) {
+        process.stderr.write(
+          `list-releases-since-tag: failed to parse gh output (${(err as Error).message})\n`,
+        );
+      }
+    } else if (ghResult.status !== 0) {
+      // Don't hard-fail — gh may not be configured in test environments.
```
Comment (mellanon) [inline scripts/list-releases-since-tag.ts]: **[major/correctness]** `gh pr list` failure silently produces an empty PR list.

When `gh` fails (rate limit, auth expiry, network blip, missing remote), the script writes a one-line stderr warning and returns with `prs: []`. Callers — `prepare-changelog.ts` in particular — can't distinguish *"no PRs merged since the last tag"* from *"gh broke"*. The release notes silently lose every PR.

This is a release-critical tool failing open. The operator running `bun scripts/prepare-changelog.ts` reads the resulting markdown, sees "_No changes._" or a partial group list, and may not connect that back to a transient `gh` failure scrolling past in stderr.

**Fix:** propagate the failure. Either throw (let the caller decide) or return a discriminated result `{ok: false, error}`. The "don't hard-fail in test environments" comment is a real concern — handle it with a `MF_ALLOW_GH_FAILURE=1` env opt-in or by injecting a fake `gh` in tests, not by burying it in production.

File: skill/Workflows/Rollback.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+1. Resolve target tag:
+   ```bash
+   gh release list --repo the-metafactory/grove --limit 5
+   ```
+   Default to the tag immediately preceding the current production tag unless the operator overrides.
+2. Check out the tag in the release worktree (never on main):
+   ```bash
+   git -C ~/Developer/release-manager-mvp checkout v<previous>
+   ```
```
Comment (mellanon) [inline skill/Workflows/Rollback.md]: **[major/correctness]** Step 2 references a directory that doesn't exist.

`git -C ~/Developer/release-manager-mvp checkout v<previous>` — what is `~/Developer/release-manager-mvp`? Day-one scope is grove-only and the deploy target is `~/Developer/grove`. The follow-up code block ("Or, for grove specifically: `git -C ~/Developer/grove ...`") is the only correct form.

A reader who copy-pastes the first block ends up checking out a tag in a repo that has no relationship to grove's deploy artifact and then runs `bunx wrangler deploy --env production` from elsewhere. Either:
- delete the `release-manager-mvp` block (grove is the only target in this scope), or
- explain what "the release worktree" is meant to be — there is no such concept introduced anywhere else in the bundle.

Right now this section actively misleads. Pick one form and ship it.

File: scripts/check-gate-table.ts [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+  let currentGate: number | undefined;
+  for (let i = 0; i < lines.length; i++) {
+    const line = lines[i] ?? "";
+    const gateHeader = line.match(/^###\s+Gate\s+(\d+):/i);
+    if (gateHeader && gateHeader[1]) {
+      currentGate = Number(gateHeader[1]);
+      continue;
+    }
+    // Phase 1 header ends Phase 0.
+    if (/^##\s+Phase\s+1:/i.test(line)) break;
```
Comment (mellanon) [inline scripts/check-gate-table.ts]: **[nit/architecture]** `Phase 1` is the only acknowledged terminator for Phase 0.

`if (/^##\s+Phase\s+1:/i.test(line)) break;` — if the SOP file is ever reorganised (a new top-level section between Phase 0 and Phase 1, the gate table moves under a different heading hierarchy, etc.) the parser silently swallows whatever comes next as gate rows. The synthetic test exercises the happy-path "Phase 1" boundary, not the "unknown next H2" case.

Cheap fix: terminate on `^##\s` (any next H2) once `currentGate` has been set at least once. The contract becomes "Phase 0 ends at the next H2", which is structurally what the SOP encodes.

**Comment by mellanon** [comment]
## Full Review: the-metafactory/release-manager#2

**Lenses applied:** CodeQuality, Security, Hardening, Architecture, EcosystemCompliance, Performance, Duplication

The MVP shape is right — six workflow MDs in the Action/Verify/Anti-pattern triplet, three bun CLIs, a structural test that locks the contract, 69 tests passing. The grove-only scope is consistently called out and the `TODO(v0.2)` markers are honest. Workflow-shape test is exactly the kind of contract-as-a-test I wanted to see. Tests pass; `bunx tsc --noEmit` is clean once `bun install` runs.

The defects are all in `check-gate-table.ts` and the chain that consumes it.

### Findings

| # | Sev | Lens | Where | What |
|---|-----|------|-------|------|
| 1 | 🔴 blocker | Correctness | `scripts/check-gate-table.ts:244` + `:180` | Empty parsed-rows result returns `allPassed: true` and exit 0. `parseGateTable` returns `[]` for any unparseable/missing/restructured table → `runChecklist([])` → `every` on `[]` returns true → `DeployStaged.md` pre-flight reads green light → prod deploys with no gates. Fail-open on the trust gate is the worst possible default. |
| 2 | 🟡 major | Correctness | `scripts/check-gate-table.ts:180` | `skipped` rows count toward `allPassed`. A milestone of all-manual cells passes the automated check identically to a milestone of all-executed-and-green cells. Distinguish `pass` from `skipped` in the green-light flag. |
| 3 | 🟡 major | Hardening | `scripts/check-gate-table.ts:155` | `bash -c <verify-cell>` runs untrusted markdown content. Today compass is internal-only so the surface is small, but the trust boundary needs to be loud — document it, and consider a regex allowlist for the verify cell. |
| 4 | 🟡 major | Correctness | `scripts/list-releases-since-tag.ts:152` | `gh pr list` failure swallowed → `prs: []` → silent empty changelog. Release-critical tool failing open. Propagate the error or gate it behind `MF_ALLOW_GH_FAILURE=1`. |
| 5 | 🟡 major | Correctness | `skill/Workflows/Rollback.md:23` | First code block references `~/Developer/release-manager-mvp` — no such directory in day-one scope. Misleading copy-paste; only the grove form is correct. |
| 6 | 🔵 nit | Architecture | `scripts/check-gate-table.ts:113` | Phase 0 termination is `^## Phase 1:` only. Tighten to "next H2 after at least one gate" to survive future reorg. |

### Key observations

1. **The gate is the whole feature.** `TrustGateCheck` is the irreversibility brake on prod deploys, and right now it has two fail-open paths (empty rows, all-skipped rows). Both feed the same `allPassed` flag that `DeployStaged` reads. Fix the green-light semantics or the rest of the safety design rests on a misleading boolean.
2. **Silent fallbacks dominate.** `gh` failure → empty PRs. Empty rows → green gate. Skipped rows → green gate. The pattern repeats — every release-critical tool is biased toward *don't interrupt the operator*. Invert it: at the gate, surprise should be loud and stop the train.
3. **The workflow MDs are clean.** `BumpVersion`, `CutRelease`, `DeployStaged`, `Announce`, `TrustGateCheck` all read well, the action/verify/anti-pattern triplet lands. Rollback is the one with a real defect.
4. **Day-one grove-only is well-marked.** TODO markers are consistent, the SKILL.md and CLAUDE.md both repeat the constraint, the structural test asserts every workflow declares it. Good discipline.
5. **What I checked.** Read every workflow MD and every script in full; ran `bun test` (69 pass); ran `bunx tsc --noEmit` (clean after install); read the existing tests; traced the `allPassed` signal from `runChecklist` → `TrustGateCheck.md` → `DeployStaged.md` pre-flight; checked that no PR labels / arc-manifest / CLAUDE.md / agents-md drift exists; no duplication across the three scripts.

### verdict: blockers=1, majors=4, nits=1 — recommend: request-changes

Fix #1 first — that's the real teeth. #2 and #3 are next, since they harden the same gate. #4 and #5 are corrections that don't compound but do mislead the operator. #6 is whenever.

(Posting as a comment because GitHub blocks self-review request-changes; the recommendation stands.)

**Comment by mellanon** [comment]
## Sweep: Echo's review — `98abb07`

All six findings addressed in a single commit on `feat/rm-001-mvp`. Day-one grove-only scope intact; no v0.2 generalization snuck in. Tests: **69 → 83 passing** (+14 regression tests, each proves the old behavior was bad AND the new one is right). `bunx tsc --noEmit` clean.

| # | Sev | Finding | Outcome |
|---|-----|---------|---------|
| 1 | 🔴 blocker | `runChecklist([])` returned `allPassed: true` (fail-open trust gate) | **A** — empty rows now `{ allPassed: false, reason: "no gates parsed — …", gates: [], manualGatesPending: 0 }`. Three regression tests cover empty rows, milestone-filter no-match, and structurally-broken markdown. |
| 2 | 🟡 major | `skipped` rows counted toward `allPassed` | **A** — `allPassed: true` now requires ≥1 executed pass AND zero fails. Skipped rows surface as `manualGatesPending` (additive JSON field). Three regression tests cover all-manual fail-closed, mixed pass+manual = green-with-pending, dry-run preservation. |
| 3 | 🟡 major | `bash -c` on operator-editable markdown | **A** — `VERIFY_COMMAND_ALLOWLIST` regex enforced before `spawnSync`; non-matching commands recorded as `fail` with the allowlist reason. Trust boundary documented at the top of `check-gate-table.ts` and in `TrustGateCheck.md`. Four regression tests cover positive prefixes, hostile payloads (`rm -rf ~`, `eval $(curl …)`, command substitution, netcat), `runGateRow` integration, and exported-RegExp introspection. |
| 4 | 🟡 major | `gh pr list` failure silently produced empty PR list | **A** — `shouldAllowGhFailure()` extracted; defaults `false` (errors propagate). Legacy permissive behavior gated behind `MF_ALLOW_GH_FAILURE=1` for test environments. Both the spawnSync-failure path and the parseGhPrList-exception path propagate by default. Two regression tests cover default-off / explicit-on. |
| 5 | 🟡 major | `Rollback.md:23` referenced `~/Developer/release-manager-mvp` | **A** — removed the misleading block; only the `~/Developer/grove` form remains, with a note that v0.2 will introduce per-repo deploy targets. |
| 6 | 🔵 nit | Phase 0 terminator was `^## Phase 1:` only | **A** — tightened to `^##\s` once ≥1 gate row has been parsed. Survives future reorg; doesn't terminate early on a leading `## Overview` preamble (regression test added). |

### JSON shape change (additive, non-breaking)

`CheckGateReport` gains two fields:

```ts
{
  gates: GateResult[],
  allPassed: boolean,
  reason?: string,           // populated when allPassed=false (or empty rows)
  manualGatesPending: number // count of skipped non-dry-run rows requiring human follow-up
}
```

`DeployStaged.md` pre-flight now reads `allPassed: true` AND `manualGatesPending: 0` (or every pending manual gate explicitly acknowledged).

### Smoke

```
$ bun scripts/check-gate-table.ts --milestone S2-22 --checklist /tmp/empty.md
{
  "gates": [],
  "allPassed": false,
  "reason": "no gates parsed — milestone table missing or table format unrecognized for milestone S2-22",
  "manualGatesPending": 0
}
exit 1
```

Re-review please — the trust-gate semantics are the load-bearing change.

File: scripts/check-gate-table.ts [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+  if (!isVerifyCommandAllowed(command)) {
+    return {
+      ...row,
+      status: "fail",
+      output:
+        "verify command not in allowlist — manual review required " +
+        "(see VERIFY_COMMAND_ALLOWLIST in scripts/check-gate-table.ts)",
+    };
+  }
+  const result = spawnSync("bash", ["-c", command], {
```
Comment (mellanon) [inline scripts/check-gate-table.ts]: **[major/hardening]** Allowlist is prefix-only — `bash -c` still executes shell metacharacter chains.

The fix at `scripts/check-gate-table.ts:198-206` and `VERIFY_COMMAND_ALLOWLIST` at L109-110 is a **prefix anchor** (`^...`). It checks the first token, then hands the *full unwrapped command string* to `spawnSync("bash", ["-c", command])` at L207. Anything an attacker appends after an allowlisted prefix is executed by bash unchanged.

Concretely, a verify cell like:

```
| G1-1 | check | `gh issue view 22 ; rm -rf ~` |
```

→ passes `isVerifyCommandAllowed` (matches `^gh\s`)
→ goes to `bash -c "gh issue view 22 ; rm -rf ~"`
→ both commands run.

The same holds for `&&`, `||`, `|`, command substitution `$(…)` / backticks (after `unwrapBacktickCommand` strips the *outer* pair, inner backticks survive), `>` / `>>` redirection, here-docs, and `eval` injected into argv. The regression tests at `__tests__/gate-check.test.ts:339-345` only cover *standalone* hostile payloads (`rm -rf ~`, `eval $(curl …)`, `nc -e`) — none of them test chaining onto a permitted prefix, which is the realistic attack vector and the exact threat Echo named in the original finding ("anyone who can edit that file can run arbitrary shell on the release operators machine").

The trust boundary docstring at L9-19 claims (1) allowlist, (2) timeout, (3) fail-closed-on-empty mitigate the threat. (1) is the load-bearing one and it doesnt. (2) doesnt bound exfil, only execution time. The defense-in-depth value of rejecting naïve cells (`rm -rf ~` standalone) is real, but the named threat is not eliminated.

**Fix options (pick one, ideally both):**

1. After allowlist match, also reject any command containing unescaped shell metacharacters: `;`, `&&`, `||`, `|`, `$(`, backtick, `>`, `<`, `>>`, `<<`, newline. Any verify cell that needs piping should be rewritten as a shell script committed to compass and invoked by name (which the allowlist *does* gate).
2. Drop `bash -c` entirely. Tokenise the command (e.g. `shell-quote.parse` or a stricter hand-rolled splitter that errors on metacharacters), then `spawnSync(argv[0], argv.slice(1), { shell: false })`. No shell, no chaining, period.

Add regression tests covering the chaining vectors above so the next sweep can prove this is closed.

**Review by mellanon — COMMENTED** [review]
## Re-review: `98abb07` — verdict

Five of six prior findings are properly closed. One — the `bash -c` trust boundary — is **not**.

| # | Sev | Status | Notes |
|---|-----|--------|-------|
| 1 | 🔴 blocker | ✅ closed | Empty rows fail-closed at L238-248 with explicit `reason`; three regression tests at `__tests__/gate-check.test.ts:211-250` lock it. |
| 2 | 🟡 major | ✅ closed | `allPassed=true` now requires ≥1 executed pass AND zero fails (L274-298); `manualGatesPending` field surfaces skipped rows; tests at L252-324 cover all-manual / mixed / dry-run. |
| 3 | 🟡 major | ❌ **not closed — still major** | Allowlist is **prefix-only**. `gh issue view 22 ; rm -rf ~` matches `^gh\s` and gets handed to `bash -c`, which runs both. Same for `&&`, `||`, `|`, `$(...)`, redirection. Tests cover standalone hostile payloads but not chaining onto an allowlisted prefix — which is the realistic attack vector and the exact threat the original finding named. See inline comment on `scripts/check-gate-table.ts:207`. |
| 4 | 🟡 major | ✅ closed | `shouldAllowGhFailure()` defaults off; both spawnSync-failure and parse-exception paths propagate (`scripts/list-releases-since-tag.ts:65-69, 158-187`). |
| 5 | 🟡 major | ✅ closed | Misleading `~/Developer/release-manager-mvp` block gone; only `~/Developer/grove` remains (`Rollback.md:22`). |
| 6 | 🔵 nit | ✅ closed | Phase 0 terminator is now `^##\s` once a row is parsed (`check-gate-table.ts:152`); preamble + reorg regression tests at L368-417. |

### What was good

- Fail-closed semantics on empty / unparseable tables are correctly the loudest possible failure now.
- The three-state passed/failed/pending split via `manualGatesPending` is the right API change — additive, documented in the doc comment, and consumed correctly by `DeployStaged.md` pre-flight (L10).
- `MF_ALLOW_GH_FAILURE` env opt-in for the legacy permissive path is exactly the right shape — default safe, escape hatch for tests, named clearly enough that nobody flips it on by accident.
- Every fix carries regression tests that prove the old behaviour was bad AND the new behaviour is right. 14 added tests; that's the bar.
- `Why:` comments next to each non-obvious change point back at this review thread, so future reviewers won't undo the fix without knowing it's load-bearing.

### What still needs work

The trust-boundary fix advertises three mitigations (allowlist, 30s timeout, fail-closed-on-empty). The allowlist is the load-bearing one, and it doesn't hold up against command chaining. The defence-in-depth value of rejecting naïve cells is real — but the threat the original finding named ("anyone who can edit `release-checklist.md` can run arbitrary shell on the release operator's machine") is not yet eliminated. The fix is ~80% there; the missing 20% is the realistic vector.

Two fix paths in the inline comment — pick one, ideally both, plus regression tests for chained vectors.

verdict: blockers=0, majors=1, nits=0 — recommend: request-changes

File: scripts/check-gate-table.ts (hunk 7 above)
Comment (mellanon) [reply scripts/check-gate-table.ts]: Closed in 3af0f20.

**Fix layout** (per your suggestion — combined both options into a single defense in depth):

1. **`SHELL_METACHARACTER_RE`** rejects any unwrapped command containing `;` `&` `|` `$` `` ` `` `>` `<` newline `\r` `(` `)` `\` before tokenization. Failure reason: `verify cell contains shell metacharacters — manual review required`. The Echo POC (`gh issue view 22 ; rm -rf ~`) is caught here.
2. **`tokenizeVerifyCommand`** — purpose-built tokenizer (no shell-quote dep, package has none). Whitespace + single/double-quoted strings only; throws on unterminated quotes or unexpected quote inside an unquoted token.
3. **`spawnSync(program, args)` directly** — `bash -c` is gone entirely. The shell is never invoked, so even if a metacharacter slipped past step 1 it would land as a literal arg, not be interpreted.
4. The first-token allowlist (`VERIFY_COMMAND_ALLOWLIST`) stays as belt-and-braces — keeps `eval` / `nc` / `python` etc. out of verify cells, but no longer the primary defense.

**Regression tests** added to `__tests__/gate-check.test.ts` (+34, total 117):
- Chained, AND, OR, pipe, dollar-paren, redirection in/out/append, heredoc-ish, backgrounded, backtick, subshell, newline, carriage-return, backslash-escape — all rejected with the metacharacter reason, none reach `spawnSync`.
- The exact Echo POC (`gh issue view 22 ; rm -rf ~`) has its own dedicated regression test that asserts `not.toContain("not in allowlist")` to confirm the metacharacter layer (not the allowlist layer) catches it.
- Tokenizer correctness tests: whitespace splitting, single/double-quote preservation, runs of whitespace, unterminated-quote rejection.
- Sanity: `echo hello` and `echo issue view 22 --json state` (multi-arg shape) still pass and execute through the new tokenizer + direct-spawn path.

Confirmed the new tests cannot pass against pre-fix code (they import `containsShellMetacharacters` and `tokenizeVerifyCommand`, which didn't exist) — the strongest possible "would have failed" demonstration. Also confirmed pre-fix `isVerifyCommandAllowed("gh issue view 22 ; rm -rf ~")` returned `true`, exactly as you described.

Trust-boundary docstring at the top of the file updated to enumerate all six layers (steps 1–6).

`bun test`: 117 pass, 0 fail. `bunx tsc --noEmit`: clean.

**Comment by mellanon** [comment]
## Re-review: `3af0f20` — verdict

The chaining vector is closed. Three independent layers — any one of them alone would block `gh issue view 22 ; rm -rf ~`; together they make the bug structurally hard to reintroduce.

### What I checked

| Layer | Mechanism | Verified at |
|---|---|---|
| 1 | `SHELL_METACHARACTER_RE` rejects `;` `&` `|` `$` `` ` `` `>` `<` `\n` `\r` `(` `)` `\` before tokenization | `scripts/check-gate-table.ts:140` — covers chain, pipe, substitution, redirection, background, subshell, newline injection, escape |
| 2 | `tokenizeVerifyCommand` — purpose-built, no shell, throws on unterminated quotes / mid-token quotes | `scripts/check-gate-table.ts:160-204` |
| 3 | `spawnSync(program, args)` — never `bash -c` | `scripts/check-gate-table.ts:341` |
| 4 | First-token allowlist still in place as belt-and-braces | `scripts/check-gate-table.ts:120` |

The exact POC from my prior comment (`gh issue view 22 ; rm -rf ~`) is regression-tested at `__tests__/gate-check.test.ts:489-498`, and the assertion `output.not.toContain("not in allowlist")` proves the metacharacter layer is what catches it — not the allowlist falling back. That's the right shape: the load-bearing defense earned the rejection.

### What's good about the fix

- **Defense ordering is right.** Metacharacter rejection runs first, so the operator gets a specific reason ("shell metacharacters — manual review required") rather than a generic "not in allowlist" — the allowlist message would have been misleading because it used to be the only line of defense, and isn't anymore.
- **Tokenizer refuses ambiguity.** Unterminated quotes and quotes inside unquoted tokens both throw, propagating to a structured `fail` rather than a "best effort" parse. Conservative is the right posture for a trust gate.
- **`bash -c` is gone for good.** Even if a future maintainer punches a hole in `SHELL_METACHARACTER_RE`, there's no shell at runtime to interpret what slips through — metacharacters become literal argv strings to whatever `program` ended up resolved.
- **Tests cover variants, not just the one I named.** Semicolon, `&&`, `||`, `|`, `$()`, `>`, `<`, `>>`, `<<`, `&`, backtick, subshell parens, `\n`, `\r`, `\` — fifteen payloads via `it.each`, each both at the regex layer and (for the load-bearing six) at the `runGateRow` integration layer. Plus tokenizer correctness on quoted strings, whitespace runs, unterminated quotes. 117 tests passing total (83 → 117, +34 in this sweep).
- **Comments point at the bug.** `Why:` blocks at `:115-124` and `:283-301` cite this review thread by date, so the next maintainer staring at the metacharacter check knows it's load-bearing rather than incidental.

### What I deliberately did not flag

- Tilde expansion (`~/path`) is no longer performed since there's no shell. Existing verify cells don't rely on it (test suite is green), and any future cell that does will fail loudly at the program-resolution step rather than silently. Documentable rather than fixable, and not worth a nit.
- Glob characters (`*`, `?`, `[`, `]`) aren't on the metacharacter list. They don't need to be — without a shell they're literal argv, not expanded.

### verdict: blockers=0, majors=0, nits=0 — recommend: merge

Day-one grove-only scope is intact; the trust gate semantics from the prior sweep (fail-closed on empty rows, manual-pending separation) all still hold. Ship it.
